src/overlays/actors/ovl_Item_B_Heart/z_item_b_heart.c

```diff
@@ -20,16 +27,94 @@ const ActorInit Item_B_Heart_InitVars = {
     (ActorFunc)ItemBHeart_Init,
     (ActorFunc)ItemBHeart_Destroy,
     (ActorFunc)ItemBHeart_Update,
-    (ActorFunc)ItemBHeart_Draw
+    (ActorFunc)ItemBHeart_Draw,
 };
-*/
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Item_B_Heart/ItemBHeart_Init.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F_DIV1000(scale, 0, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneForward, 4000, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneScale, 800, ICHAIN_CONTINUE),
+    ICHAIN_F32(uncullZoneDownward, 800, ICHAIN_STOP),
+};
+
+void ItemBHeart_Init(Actor* thisx, PlayState* play) {
+    ItemBHeart* this = THIS;
+
+    if (Flags_GetCollectible(play, 0x1F)) {
+        Actor_Kill(&this->actor);
+        return;
+    }
+    Actor_ProcessInitChain(&this->actor, sInitChain);
+    ActorShape_Init(&this->actor.shape, 0.0f, NULL, 0.8f);
+    if (this->actor.params == BHEART_PARAM_SMALL) {
+        this->scale = BHEART_SCALE_SMALL;
+    } else {
+        this->scale = BHEART_SCALE_NORMAL;
+    }
+    this->actor.world.pos.y += 20.0f * this->scale;
+}
+
+void ItemBHeart_Destroy(Actor* thisx, PlayState* play) {
+}
+
+/**
+ * Adjusts size and handles collection (if of proper scale)
+ */
+void ItemBHeart_Update(Actor* thisx, PlayState* play) {
+    ItemBHeart* this = THIS;
+
+    ItemBHeart_UpdateRotationAndScale(this, play);
+
+    if (!(this->scale < BHEART_SCALE_MIN_COLLECTIBLE)) {
+        if (Actor_HasParent(&this->actor, play)) {
+            Flags_SetCollectible(play, 0x1F);
+            Actor_Kill(&this->actor);
+            return;
+        }
+        Actor_OfferGetItem(&this->actor, play, GI_HEART_CONTAINER, 30.0f, 80.0f);
+    }
+}
+
+/**
+ * Rotate continuously while approaching 40% of object's unit scale.
+ */
+void ItemBHeart_UpdateRotationAndScale(ItemBHeart* this, PlayState* play) {
```

Comment:
Thoughts on maybe just `ItemBHeart_UpdateModel`? Feel like it still describes it pretty well, and is more succinct.